subprojects/execution/src/main/java/org/gradle/internal/execution/steps/BuildCacheStep.java

```diff
@@ -149,11 +148,11 @@ public void visitLocalState(File localStateRoot) {
         });
     }
 
-    private CurrentSnapshotResult executeAndStoreInCache(CacheableWork work, BuildCacheKey cacheKey, IncrementalChangesContext context) {
-        CurrentSnapshotResult result = executeWithoutCache(context);
+    private CurrentSnapshotResult executeAndStoreInCache(CacheableWork cacheableWork, BuildCacheKey cacheKey, IncrementalChangesContext context) {
+        CurrentSnapshotResult result = executeWithoutCache(cacheableWork.work, context);
```

Comment:
It seems better to me to pass the unit of work separately from the cacheable entity and not access the unit of work on the implementation of the cacheable entity. One way to do that would be to create the cacheable entity from the work in `executeAndStoreInCache` instead of passing it around. Or add a `getCacheableEntity()` to `IncrementalChangesContext`.